Add type hints

dcermak opened the pull request:
Resurrection of #18, but includes the changes from #22 as well, as the CI will otherwise fail.

AdamWill commented on tests/test_client.py:
```diff
@@ -97,7 +97,7 @@ def test_add_auth_headers(self, simple_config):
         assert prepared.headers != authed.headers
         assert authed.headers["X-API-Hash"] == "71373f0a57118b120d1915ccc0a24ae2cc112ad3"
-        assert authed.headers["X-API-Microtime"] == b"1582761600.0"
+        assert authed.headers["X-API-Microtime"] == "1582761600.0"
```
is this...okay? Are we sure this works? I'm pretty sure I wrote this to behave the same as the perl client, so this just worries me a bit.

dcermak replied:
I am honestly not sure, @okurz do you have an idea? The issue is that requests' typing complains that `bytes` are not supposed to be in the header. But that could be wrong of course.

okurz replied:
I don't know. But can you exclude this from the typing check somehow?

AdamWill commented:
I merged #22 , so can you rebase this? Thanks.

I really hate the `@overload` stuff, just visually, but I don't see a way to lose it while maintaining compatibility sadly...

dcermak commented:
> I merged #22 , so can you rebase this? Thanks.

Done

> I really hate the `@overload` stuff, just visually, but I don't see a way to lose it while maintaining compatibility sadly...

The `@overload` stuff is not strictly necessary, the type hints work without it. It just provides the additional benefit that the type checker can infer more precise return types if it knows some parameters "at compile time" (i.e. when scanning your code). E.g. if you write `do_request(…, parse=False)`, mypy will now know that the return type is a `requests.Response`. If we omit the overloads, then it would only know that it's either a `requests.Response` or `Any`. It's not overly useful in this case, so if it bothers you, we can remove it entirely.

AdamWill commented:
I merged some other PRs, so can you rebase this again? Sorry. I want to check it out and see if authed requests actually work with it before merging, also check the tests still pass. Note, I've renamed the default branch to 'main' so you'll have to change that on your local checkout.

okurz commented:
@Mergifyio rebase

dcermak commented:
@AdamWill done. I also had to revert https://github.com/dcermak/openQA-python-client/commit/f2db9d08dacefd3581f5c546e5c4af1f158d9fbd as that set the comparison branch to `origin/master` which is now gone.

AdamWill commented:
@dcermak  It just needs to be changed to origin/main. Sorry, thought I did that, I'll do it on main branch now and we can rebase this again...

AdamWill commented:
Done, can you rebase again and leave `tox.ini` alone, hopefully it should work? Thanks.

dcermak commented:
Adam Williamson ***@***.***> writes:

> Done, can you rebase again and leave `tox.ini` alone, hopefully it should work? Thanks.

I have dropped by revert and rebased on main.
